src/wagtail_vector_index/index/models.py

```diff
@@ -0,0 +1,402 @@
+            yield from self.to_documents(object, embedding_backend=embedding_backend)
+
+    def bulk_from_documents(
+        self, documents: Iterable[Document]
+    ) -> Generator[models.Model, None, None]:
+        # TODO: Implement a more efficient approach
+        for document in documents:
+            yield self.from_document(document)
+
+
+class ModelVectorIndex(VectorIndex):
+    """A VectorIndex which indexes the results of querysets of EmbeddableFieldsMixin models"""
+
+    querysets: ClassVar[Sequence[models.QuerySet]]
+    converter_class: ClassVar[
+        type[EmbeddableFieldsDocumentConverter]
+    ] = EmbeddableFieldsDocumentConverter
+
+    def _get_querysets(self) -> Sequence[models.QuerySet]:
+        return self.querysets
+
+    def get_converter(self) -> EmbeddableFieldsDocumentConverter:
+        queryset_models = [qs.model for qs in self._get_querysets()]
+        all_the_same = len(set(queryset_models)) == 1
+        if not all_the_same:
+            raise ValueError(
+                "All querysets must be of the same model to use the default converter."
+            )
+        return self.converter_class(queryset_models[0])
+
+    def rebuild_index(self):
+        """Before building an index, generate Embedding objects for everything in
+        this index"""
+        querysets = self._get_querysets()
+
+        for queryset in querysets:
+            for instance in queryset:
+                self.get_converter().generate_embeddings(
+                    instance, embedding_backend=self.embedding_backend
+                )
+
+        return super().rebuild_index()
+
+    def get_documents(self) -> Iterable[Document]:
+        querysets = self._get_querysets()
+
+        for queryset in querysets:
+            instances = queryset.prefetch_related("embeddings")
+            yield from self.get_converter().bulk_to_documents(
+                instances, embedding_backend=self.embedding_backend
+            )
+
+
+class PageVectorIndex(ModelVectorIndex):
+    """A model vector indexed for use with Wagtail pages that automatically
+    restricts indexed models to live pages."""
+
+    querysets: Sequence[PageQuerySet]
+
+    def _get_querysets(self) -> list[PageQuerySet]:
+        qs_list = super()._get_querysets()
+
+        # Technically a manager instance, not a queryset, but we want to use the custom
+        # methods.
+        return [cast(PageQuerySet, qs).live() for qs in qs_list]
+
+
+# ###########
+# Classes related to automatic generation of indexes for models
+# ###########
+
+
+class GeneratedIndexMixin(models.Model):
+    """Mixin for Django models that automatically generates registers a VectorIndex for the model"""
+
+    vector_index_class: ClassVar[Optional[type[VectorIndex]]] = None
+    converter_class: ClassVar[type[DocumentConverter]]
+
+    class Meta:
+        abstract = True
+
+    @classmethod
+    def get_vector_index(cls):
+        """Get a vector index instance for this model"""
```

Comment:
**Suggestion:** Could the comment elaborate that this index is only the auto-generated one? I know that for the niceness of the method name, it shouldn't probably say something like `get_default_vector_index` or `get_autogenerated_vector_index`. I guess we should establish a project glossary (could add a documentation page for this) where we decide to stick to one adjective for this type of default auto-generated model vector index.